src/run/runner/wall_time/perf/debug_info.rs

```diff
@@ -0,0 +1,206 @@
+            ))
+        };
+
+        let dwarf = gimli::Dwarf::load(load_section)?;
+        addr2line::Context::from_dwarf(dwarf).map_err(Into::into)
+    }
+}
+
+impl ModuleDebugInfoExt for ModuleDebugInfo {
+    /// Create debug info from existing symbols by looking up file/line in DWARF
+    fn from_symbols<P: AsRef<Path>>(path: P, symbols: &ModuleSymbols) -> anyhow::Result<Self> {
+        let content = std::fs::read(path.as_ref())?;
+        let object = object::File::parse(&*content)?;
+
+        let ctx = Self::create_dwarf_context(&object).context("Failed to create DWARF context")?;
+        let load_bias = symbols.load_bias();
+        let (mut min_addr, mut max_addr) = (u64::MAX, 0);
+        let debug_infos = symbols
+            .symbols()
+            .iter()
+            .filter_map(|symbol| {
+                let (file, line) = match ctx.find_location(symbol.addr) {
+                    Ok(Some(location)) => {
+                        let file = location.file.map(|f| f.to_string())?;
+                        (file, location.line)
+                    }
+                    _ => return None,
+                };
+
+                min_addr = min_addr.min(symbol.addr);
+                max_addr = max_addr.max(symbol.addr + symbol.size);
+
+                Some(DebugInfo {
+                    addr: symbol.addr,
+                    size: symbol.size,
+                    name: symbol.name.clone(),
+                    file,
+                    line,
+                })
+            })
+            // Sort by address, to allow binary search lookups in backend
+            .sorted_by_key(|d| d.addr)
+            .collect();
+
+        Ok(ModuleDebugInfo {
+            object_path: path.as_ref().to_string_lossy().to_string(),
+            load_bias,
+            addr_bounds: (min_addr, max_addr),
+            debug_infos,
+        })
+    }
+}
+
+/// Represents all the modules inside a process and their debug info.
+pub struct ProcessDebugInfo {
+    pid: pid_t,
+    modules: Vec<ModuleDebugInfo>,
+}
+
+impl ProcessDebugInfo {
+    pub fn new(
+        pid: pid_t,
+        process_symbols: &crate::run::runner::wall_time::perf::perf_map::ProcessSymbols,
+    ) -> Self {
+        let mut modules = Vec::new();
+        for (path, module_symbols) in process_symbols.modules_with_symbols() {
+            match ModuleDebugInfo::from_symbols(path, module_symbols) {
+                Ok(module_debug_info) => {
+                    modules.push(module_debug_info);
+                }
+                Err(error) => {
+                    debug!("Failed to load debug info for module {path:?}: {error}");
+                }
+            }
+        }
+
+        Self { pid, modules }
+    }
+
+    pub fn save_to<P: AsRef<std::path::Path>>(&self, folder: P) -> anyhow::Result<()> {
+        if self.modules.is_empty() {
+            return Ok(());
+        }
+
+        let debug_info_path = folder.as_ref().join(format!("{}.debuginfo", self.pid));
```

Comment:
As discussed, we should put this into the perf.metadata or something, in order to avoid creating another file format that we need to maintain.